pkg/kubelet/container/cache.go

```diff
@@ -0,0 +1,187 @@
+	pods map[types.UID]*data
+	// A global timestamp represents how fresh the cached data is. All
+	// cache content is at the least newer than this timestamp. Note that the
+	// timestamp is nil after initialization, and will only become non-nil when
+	// it is ready to serve the cached statuses.
+	timestamp *time.Time
+	// Map that stores the callback records.
+	callbacks map[types.UID]callbackRecord
+}
+
+// NewCache creates a pod cache.
+func NewCache() Cache {
+	return &cache{pods: map[types.UID]*data{}, callbacks: map[types.UID]callbackRecord{}}
+}
+
+// getIfNewerThan returns the data it is newer than the given time.
+// Otherwise, it returns nil.
+func (c *cache) getIfNewerThan(id types.UID, minTime time.Time) *data {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	d, ok := c.pods[id]
+
+	if !ok {
+		// Status is not cached.
+		if c.timestamp != nil && c.timestamp.After(minTime) {
+			// The global timestamp is newer than minTime, return status.
+			return &data{status: &PodStatus{ID: id}, err: nil}
+		}
+	} else {
+		// Status is cached, return status if either of the following is true.
+		//   * status was modified after minTime
+		//   * the global timestamp of the cache is newer than minTime.
+		if d.modified.After(minTime) || c.timestamp.After(minTime) {
+			return d
+		}
+	}
+
+	// The pod status is not ready.
+	return nil
+}
+
+// Get returns the PodStatus for the pod; callers are expected not to
+// modify the objects returned.
+func (c *cache) Get(id types.UID) (*PodStatus, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	d := c.get(id)
+	return d.status, d.err
+}
+
+func (c *cache) get(id types.UID) *data {
+	d, ok := c.pods[id]
+	if !ok {
+		// A cache miss means that there are no states regarding the pod when
+		// querying the container runtime; return an empty PodStatus.
+		return &data{status: &PodStatus{ID: id}, err: nil}
+	}
+	return d
+}
+
+// Set sets the PodStatus for the pod.
+func (c *cache) Set(id types.UID, status *PodStatus, err error, timestamp time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	record, ok := c.callbacks[id]
+	if ok {
+		defer func() {
+			nonblockingSend(record.ch, c.get(id))
+			delete(c.callbacks, id)
+		}()
+	}
+
+	if err == nil && len(status.ContainerStatuses) == 0 {
+		// Remove if status is empty. This allows the cache to reclaims the
+		// unused entries without needing explicit garbage collection methods.
+		delete(c.pods, id)
+		return
+	}
+	c.pods[id] = &data{status: status, err: err, modified: timestamp}
+}
+
+// SetTimestamp modifies the global timestamp of the cache.
+func (c *cache) SetGlobalTimestamp(timestamp time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.timestamp = &timestamp
+	c.notify()
+}
+
+func nonblockingSend(ch chan *data, d *data) {
+	select {
+	case ch <- d:
+	default:
+		// Don't block.
+	}
+}
+
+func (c *cache) notify() {
+	for id, r := range c.callbacks {
+		if c.timestamp.After(r.time) {
+			nonblockingSend(r.ch, c.get(id))
```

Comment:
This doesn't really delete the channel, but just delete the callback record in the map. Since `GetNewerThan()` always creates a new channel, it should not get stuck as long as there is a notification inflight. I'll close the channel just in case, but there is nothing the goroutine can do if the channel gets closed without any preceding notification...